src_c/rect.c

```diff
@@ -695,6 +695,132 @@ pg_rect_collidelistall(pgRectObject *self, PyObject *args)
     return ret;
 }
 
+static SDL_Rect *
+pgRect_FromObjectAndKeyFunc(PyObject *obj, PyObject *keyfunc)
+{
+    SDL_Rect temp;
+    if (keyfunc == NULL) {
+        return pgRect_FromObject(obj, &temp);
+    }
+
+    PyObject *obj_with_rect = PyObject_CallFunctionObjArgs(keyfunc, obj, NULL);
+    if (obj_with_rect == NULL) {
+        return NULL;
+    }
+
+    return pgRect_FromObject(obj_with_rect, &temp);
+}
+
+static PyObject *
+pg_rect_collideobjectsall(pgRectObject *self, PyObject *args, PyObject *kwargs)
+{
+    SDL_Rect *argrect;
+    Py_ssize_t size;
+    int loop;
+    PyObject *list, *obj;
+    PyObject *keyfunc = NULL;
+    PyObject *ret = NULL;
+    static char *keywords[] = {"list", "key", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "O|$O:collideobjectsall",
+                                     keywords, &list,
+                                     &keyfunc)) {
+        return RAISE(PyExc_TypeError,
+                     "Could not parse args and kw.");
+    }
+
+    if (!PySequence_Check(list)) {
+        return RAISE(PyExc_TypeError,
+                     "Argument must be a sequence of objects.");
+    }
+
+    if (keyfunc == Py_None) {
+        keyfunc = NULL;
+    }
+
+    if (keyfunc != NULL  && !PyCallable_Check(keyfunc)) {
+        return RAISE(PyExc_TypeError,
+                     "Key function must be callable with one argument.");
+    }
+
+    ret = PyList_New(0);
+    if (!ret) {
+        return NULL;
+    }
+
+    size = PySequence_Length(list); /*warning, size could be -1?*/
+    for (loop = 0; loop < size; ++loop) {
```

Comment:
Please keep a good eyes on the ref counting, not sure if I did it correctly everywhere (executing this method in a loop did not reveal any memory increase as a simple test... but that might not be very accurate)

Reply:
There don't seem to be memleaks here, but there are error handling bugs, which I have reviewed below